fido2/ctap.c

```diff
@@ -1154,6 +1157,270 @@ uint8_t ctap_get_next_assertion(CborEncoder * encoder)
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&rp, "id");
+        check_ret(ret);
+        if (memcmp(ssh_hash, rk.id.rpIdHash, 32) == 0) {
+            // recover the rpId from the hash because Solo stores only the hash :(
+            ret = cbor_encode_text_stringz(&rp, "ssh:");
+        } else {
+            ret = cbor_encode_text_stringz(&rp, "");
+        }
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&rp, "name");
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&rp, rk.user.name);
+        check_ret(ret);
+        ret = cbor_encoder_close_container(&map, &rp);
+        check_ret(ret);
+    }
+    ret = cbor_encode_int(&map, 4);
+    check_ret(ret);
+    cbor_encode_byte_string(&map, rk.id.rpIdHash, 32);
+    check_ret(ret);
+    if (rp_count > 0)
+    {
+        ret = cbor_encode_int(&map, 5);
+        check_ret(ret);
+        ret = cbor_encode_int(&map, rp_count);
+        check_ret(ret);
+    }
+    ret = cbor_encoder_close_container(encoder, &map);
+    check_ret(ret);
+    return 0;
+}
+
+uint8_t ctap_cred_rk(CborEncoder * encoder, int rk_ind, int rk_count)
+{
+    CTAP_residentKey rk;
+    ctap_load_rk(rk_ind, &rk);
+
+    CborEncoder map;
+    size_t map_size = rk_count > 0 ? 4 : 3;
+    int ret = cbor_encoder_create_map(encoder, &map, map_size);
+    check_ret(ret);
+    ret = cbor_encode_int(&map, 6);
+    check_ret(ret);
+    {
+        CborEncoder usr;
+        ret = cbor_encoder_create_map(&map, &usr, 4);
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, "id");
+        check_ret(ret);
+        ret = cbor_encode_byte_string(&usr, rk.user.id, rk.user.id_size);
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, "icon");
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, rk.user.icon);
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, "name");
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, rk.user.name);
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, "displayName");
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, rk.user.displayName);
+        check_ret(ret);
+        ret = cbor_encoder_close_container(&map, &usr);
+        check_ret(ret);
+    }
+
+    ret = cbor_encode_int(&map, 7);
+    check_ret(ret);
+    {
+        CborEncoder credId;
+        ret = cbor_encoder_create_map(&map, &credId, 1);
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&credId, "id");
+        check_ret(ret);
+        ret = cbor_encode_byte_string(&credId, (uint8_t*)&rk.id, sizeof(CredentialId));
+        check_ret(ret);
+        ret = cbor_encoder_close_container(&map, &credId);
+        check_ret(ret);
+    }
+
+    ret = cbor_encode_int(&map, 8);
+    check_ret(ret);
+    ctap_generate_cose_key(&map, (uint8_t*)&rk.id, sizeof(CredentialId), PUB_KEY_CRED_PUB_KEY, COSE_ALG_ES256);
+
+    if (rk_count > 0)
```

Comment:
`if (rk_count > 0 && rk_ind == 0)`

Reply:
`CM_cmdRKBegin` doesn't always start with the first resident key. We may have to skip several RKs until we reach the one with the specified hash.

`rk_count` is greater than zero only for `CM_cmdRKBegin`, see line 1380

Reply:
Gotcha, thanks :)